File: app/src/main/java/de/westnordost/streetcomplete/quests/bikeway/AddCyclewayForm.java
package de.westnordost.streetcomplete.quests.bikeway;

import android.os.Bundle;
import android.support.annotation.AnyThread;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Map;

import de.westnordost.streetcomplete.R;
import de.westnordost.streetcomplete.data.osm.tql.FiltersParser;
import de.westnordost.streetcomplete.data.osm.tql.TagFilterExpression;
import de.westnordost.streetcomplete.quests.AbstractQuestFormAnswerFragment;
import de.westnordost.streetcomplete.quests.StreetSideRotater;
import de.westnordost.streetcomplete.view.ListAdapter;
import de.westnordost.streetcomplete.view.StreetSideSelectPuzzle;


public class AddCyclewayForm extends AbstractQuestFormAnswerFragment
{
	public static final String
			CYCLEWAY_LEFT = "cycleway_left",
			CYCLEWAY_RIGHT = "cycleway_right",
			CYCLEWAY_LEFT_DIR = "cycleway_left_opposite",
			CYCLEWAY_RIGHT_DIR = "cycleway_right_opposite",
			IS_ONEWAY_NOT_FOR_CYCLISTS = "oneway_not_for_cyclists";

	private static final String
			DEFINE_BOTH_SIDES = "define_both_sides";

	private static final TagFilterExpression LIKELY_NO_BICYCLE_CONTRAFLOW = new FiltersParser().parse(
			"ways with oneway:bicycle != no and " +
			" (oneway ~ yes|-1 and highway ~ primary|secondary|tertiary or junction=roundabout)");


	private StreetSideRotater streetSideRotater;
	StreetSideSelectPuzzle puzzle;

	private boolean isDefiningBothSides;

	private Cycleway leftSide;
	private Cycleway rightSide;

	@Override public View onCreateView(LayoutInflater inflater, ViewGroup container,
									   Bundle inState)
	{
		View view = super.onCreateView(inflater, container, inState);
		setContentView(R.layout.quest_street_side_puzzle);
		setNoContentPadding();

		View compassNeedle = view.findViewById(R.id.compassNeedle);

		puzzle = view.findViewById(R.id.puzzle);
		puzzle.setListener(this::showCyclewaySelectionDialog);

		streetSideRotater = new StreetSideRotater(puzzle, compassNeedle, getElementGeometry());

		initPuzzleDisplay(inState);
		initPuzzleImages(inState);

		return view;
	}

	private void initPuzzleDisplay(Bundle inState)
	{
		if(inState != null)
		{
			isDefiningBothSides = inState.getBoolean(DEFINE_BOTH_SIDES);
		}
		else
		{
			isDefiningBothSides = !LIKELY_NO_BICYCLE_CONTRAFLOW.matches(getOsmElement());
		}

		if(!isDefiningBothSides)
		{
			if(isLeftHandTraffic()) puzzle.showOnlyLeftSide();
			else                    puzzle.showOnlyRightSide();

			if(!"roundabout".equals(getOsmElement().getTags().get("junction")))
			{
				addOtherAnswer(R.string.quest_cycleway_answer_contraflow_cycleway, this::showBothSides);
			}
		}
	}

	private void initPuzzleImages(Bundle inState)
	{
		int defaultResId = isLeftHandTraffic() ?
				R.drawable.ic_cycleway_unknown_l : R.drawable.ic_cycleway_unknown;

		if(inState != null)
		{
			String rightSideString = inState.getString(CYCLEWAY_RIGHT);
			if(rightSideString != null)
			{
				rightSide = Cycleway.valueOf(rightSideString);
				checkIsFormComplete();
				puzzle.setRightSideImageResource(rightSide.getIconResId(isLeftHandTraffic()));
			}
			else
			{
				puzzle.setRightSideImageResource(defaultResId);
			}
			String leftSideString = inState.getString(CYCLEWAY_LEFT);
			if(leftSideString != null)
			{
				leftSide = Cycleway.valueOf(leftSideString);
				checkIsFormComplete();
				puzzle.setLeftSideImageResource(leftSide.getIconResId(isLeftHandTraffic()));
			}
			else
			{
				puzzle.setLeftSideImageResource(defaultResId);
			}
		}
		else
		{
			puzzle.setLeftSideImageResource(defaultResId);
			puzzle.setRightSideImageResource(defaultResId);
		}
	}

	@Override public void onSaveInstanceState(Bundle outState)
	{
		super.onSaveInstanceState(outState);
		if(rightSide != null) outState.putString(CYCLEWAY_RIGHT, rightSide.name());
		if(leftSide != null)  outState.putString(CYCLEWAY_LEFT, leftSide.name());
		outState.putBoolean(DEFINE_BOTH_SIDES, isDefiningBothSides);
	}

	@AnyThread public void onMapOrientation(final float rotation, final float tilt)
	{
		if(streetSideRotater != null) {
			streetSideRotater.onMapOrientation(rotation, tilt);
		}
	}

	@Override protected void onClickOk()
	{
		boolean isOnewayNotForCyclists = false;

		// a cycleway that goes into opposite direction of a oneway street needs special tagging
		Bundle bundle = new Bundle();
		if(isOneway() && leftSide != null && rightSide != null)
		{
			// if the road is oneway=-1, a cycleway that goes opposite to it would be cycleway:oneway=yes
			int reverseDir = isReversedOneway() ? 1 : -1;

			if(isReverseSideRight())
			{
				if(isSingleTrackOrLane(rightSide))
				{
					bundle.putInt(CYCLEWAY_RIGHT_DIR, reverseDir);
				}
				isOnewayNotForCyclists = rightSide != Cycleway.NONE;
			}
			else
			{
				if(isSingleTrackOrLane(leftSide))
				{
					bundle.putInt(CYCLEWAY_LEFT_DIR, reverseDir);
				}
				isOnewayNotForCyclists = leftSide != Cycleway.NONE;
			}

			isOnewayNotForCyclists |= isDualTrackOrLane(leftSide);
			isOnewayNotForCyclists |= isDualTrackOrLane(rightSide);
		}

		if(leftSide != null)  bundle.putString(CYCLEWAY_LEFT, leftSide.name());
		if(rightSide != null) bundle.putString(CYCLEWAY_RIGHT, rightSide.name());
		bundle.putBoolean(IS_ONEWAY_NOT_FOR_CYCLISTS, isOnewayNotForCyclists);
		applyAnswer(bundle);
	}

	private static boolean isSingleTrackOrLane(Cycleway cycleway)
	{
		return cycleway == Cycleway.TRACK || cycleway == Cycleway.EXCLUSIVE_LANE;
	}

	private static boolean isDualTrackOrLane(Cycleway cycleway)
	{
		return cycleway == Cycleway.DUAL_TRACK || cycleway == Cycleway.DUAL_LANE;
	}

	@Override public boolean isFormComplete()
	{
		if(isDefiningBothSides) return leftSide != null && rightSide != null;
		else return leftSide != null || rightSide != null;
	}

	private void showCyclewaySelectionDialog(final boolean isRight)
	{
		RecyclerView recyclerView = new RecyclerView(getActivity());
		recyclerView.setLayoutParams(new RecyclerView.LayoutParams(
				ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT));
		recyclerView.setLayoutManager(new GridLayoutManager(getActivity(), 2));

		final AlertDialog alertDialog = new AlertDialog.Builder(getActivity())
				.setTitle(R.string.quest_select_hint)
				.setView(recyclerView)
				.create();

		recyclerView.setAdapter(createAdapter(getCyclewayItems(isRight), cycleway ->
		{
			alertDialog.dismiss();

			int iconResId = cycleway.getIconResId(isLeftHandTraffic());

			if (isRight)
			{
				puzzle.replaceRightSideImageResource(iconResId);
				rightSide = cycleway;
			}
			else
			{
				puzzle.replaceLeftSideImageResource(iconResId);
				leftSide = cycleway;
			}
			checkIsFormComplete();
		}));
		alertDialog.show();
	}

	private List<Cycleway> getCyclewayItems(boolean isRight)
	{
		List<Cycleway> values = new ArrayList<>(Arrays.asList(Cycleway.getDisplayValues()));
		// different wording for a contraflow lane that is marked like a "shared" lane (just bicycle pictogram)
		if(isOneway() && isReverseSideRight() == isRight)
		{
			Collections.replaceAll(values, Cycleway.PICTOGRAMS, Cycleway.NONE_NO_ONEWAY);
		}
		String country = getCountryInfo().getCountryCode();
		if("BE".equals(country))
		{
			// Belgium does not make a difference between continuous and dashed lanes -> so don't tag that difference
			// also, in Belgium there is a differentiation between the normal lanes and suggestion lanes
			values.remove(Cycleway.EXCLUSIVE_LANE);
			values.remove(Cycleway.ADVISORY_LANE);
			values.add(0, Cycleway.LANE_UNSPECIFIED);
			values.add(1, Cycleway.SUGGESTION_LANE);
		}
		else if("NL".equals(country))
		{
			// a differentiation between dashed lanes and suggestion lanes only exist in NL and BE
			values.add(values.indexOf(Cycleway.ADVISORY_LANE)+1, Cycleway.SUGGESTION_LANE);
		}

		return values;
	}

	private interface OnCyclewaySelected { void onCyclewaySelected(Cycleway cycleway); }
	private ListAdapter<Cycleway> createAdapter(List<Cycleway> items, final OnCyclewaySelected callback)
	{
		return new ListAdapter<Cycleway>(items)
		{
			@Override public ViewHolder<Cycleway> onCreateViewHolder(ViewGroup parent, int viewType)
			{
				return new ViewHolder<Cycleway>(LayoutInflater.from(parent.getContext()).inflate(
						R.layout.labeled_icon_button_cell, parent, false))
				{
					@Override protected void onBind(final Cycleway item)
					{
						ImageView iconView = itemView.findViewById(R.id.imageView);
						TextView textView = itemView.findViewById(R.id.textView);
						int resId = item.getIconResId(isLeftHandTraffic());
						iconView.setImageDrawable(getResources().getDrawable(resId));
						textView.setText(item.nameResId);
						itemView.setOnClickListener(view -> callback.onCyclewaySelected(item));
					}
				};
			}
		};
	}

	private void showBothSides()
	{
		isDefiningBothSides = true;
		puzzle.showBothSides();
	}

	private boolean isOneway()
	{
		Map<String, String> tags = getOsmElement().getTags();
		String oneway = tags.get("oneway");
		return oneway != null && (oneway.equals("yes") || oneway.equals("-1"));
	}

	/** @return whether the side that goes into the opposite direction of the driving direction of a
	 *          one-way is on the right side of the way */
	private boolean isReverseSideRight()
	{
		return isReversedOneway() ^ isLeftHandTraffic();
	}

	private boolean isReversedOneway()
	{
		return "-1".equals(getOsmElement().getTags().get("oneway"));
	}

	// just a shortcut
	private boolean isLeftHandTraffic()
	{
		return getCountryInfo().isLeftHandTraffic();
	}
}

File: app/src/main/java/de/westnordost/streetcomplete/quests/bikeway/AddCyclewayForm.kt
package de.westnordost.streetcomplete.quests.bikeway

import android.os.Bundle
import androidx.annotation.AnyThread
import android.view.View

import java.util.Collections

import de.westnordost.streetcomplete.R
import de.westnordost.streetcomplete.data.osm.elementgeometry.ElementPolylinesGeometry
import de.westnordost.streetcomplete.data.elementfilter.ElementFiltersParser
import de.westnordost.streetcomplete.quests.AbstractQuestFormAnswerFragment
import de.westnordost.streetcomplete.quests.OtherAnswer
import de.westnordost.streetcomplete.quests.StreetSideRotater
import de.westnordost.streetcomplete.view.image_select.ImageListPickerDialog
import kotlinx.android.synthetic.main.quest_street_side_puzzle.*
import kotlinx.android.synthetic.main.view_little_compass.*

class AddCyclewayForm : AbstractQuestFormAnswerFragment<CyclewayAnswer>() {

    override val contentLayoutResId = R.layout.quest_street_side_puzzle
    override val contentPadding = false

    private var isDisplayingPreviousCycleway: Boolean = false

    override val otherAnswers: List<OtherAnswer> get() {
        val isNoRoundabout = osmElement!!.tags["junction"] != "roundabout"
        val result = mutableListOf<OtherAnswer>()
        if (!isDefiningBothSides && isNoRoundabout) {
            result.add(OtherAnswer(R.string.quest_cycleway_answer_contraflow_cycleway) { showBothSides() })
        }
        return result
    }

    private val likelyNoBicycleContraflow = ElementFiltersParser().parse("""
            ways with oneway:bicycle != no and (
                oneway ~ yes|-1 and highway ~ primary|primary_link|secondary|secondary_link|tertiary|tertiary_link|unclassified
                or junction = roundabout
            )
        """)

    private var streetSideRotater: StreetSideRotater? = null

    private var isDefiningBothSides: Boolean = false

    private var leftSide: Cycleway? = null
    private var rightSide: Cycleway? = null

    /** returns whether the side that goes into the opposite direction of the driving direction of a
     * one-way is on the right side of the way */
    private val isReverseSideRight get() = isReversedOneway xor isLeftHandTraffic

    private val isOneway get() = isForwardOneway || isReversedOneway

    private val isForwardOneway get() = osmElement!!.tags["oneway"] == "yes"
    private val isReversedOneway get() = osmElement!!.tags["oneway"] == "-1"

    // just a shortcut
    private val isLeftHandTraffic get() = countryInfo.isLeftHandTraffic

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        if (savedInstanceState == null) {
            initStateFromTags()
        } else {
            onLoadInstanceState(savedInstanceState)
        }

        puzzleView.listener = { isRight -> showCyclewaySelectionDialog(isRight) }

        streetSideRotater = StreetSideRotater(puzzleView, compassNeedleView, elementGeometry as ElementPolylinesGeometry)

        if (!isDefiningBothSides) {
            if (isLeftHandTraffic) puzzleView.showOnlyLeftSide()
            else                   puzzleView.showOnlyRightSide()
        }

        val defaultResId =
            if (isLeftHandTraffic) R.drawable.ic_cycleway_unknown_l
            else                   R.drawable.ic_cycleway_unknown

        val defaultTitleId = R.string.quest_street_side_puzzle_select

        puzzleView.setLeftSideImageResource(leftSide?.getIconResId(isLeftHandTraffic) ?: defaultResId)
        puzzleView.setRightSideImageResource(rightSide?.getIconResId(isLeftHandTraffic) ?: defaultResId)
        puzzleView.setLeftSideText(resources.getString(leftSide?.getTitleResId() ?: defaultTitleId ))
        puzzleView.setRightSideText(resources.getString(rightSide?.getTitleResId() ?: defaultTitleId ))

        checkIsFormComplete()
    }

    private fun initStateFromTags() {
        val countryCode = countryInfo.countryCode
        val sides = createCyclewaySides(osmElement!!.tags, isLeftHandTraffic)
        val left = sides?.left?.takeIf { it.isAvailableAsSelection(countryCode) }
        val right = sides?.right?.takeIf { it.isAvailableAsSelection(countryCode) }
        val bothSidesWereDefinedBefore = sides?.left != null && sides.right != null

        leftSide = left
        rightSide = right
        isDefiningBothSides = bothSidesWereDefinedBefore || !likelyNoBicycleContraflow.matches(osmElement!!)

        // only show as re-survey (yes/no button) if the previous tagging was complete
        setAsResurvey(isFormComplete())
    }

    private fun onLoadInstanceState(savedInstanceState: Bundle) {
        isDefiningBothSides = savedInstanceState.getBoolean(DEFINE_BOTH_SIDES)
        savedInstanceState.getString(CYCLEWAY_RIGHT)?.let { rightSide = Cycleway.valueOf(it) }
        savedInstanceState.getString(CYCLEWAY_LEFT)?.let { leftSide = Cycleway.valueOf(it) }
        isDisplayingPreviousCycleway = savedInstanceState.getBoolean(IS_DISPLAYING_PREVIOUS_CYCLEWAY)

    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        rightSide?.let { outState.putString(CYCLEWAY_RIGHT, it.name) }
        leftSide?.let { outState.putString(CYCLEWAY_LEFT, it.name) }
        outState.putBoolean(DEFINE_BOTH_SIDES, isDefiningBothSides)
        outState.putBoolean(IS_DISPLAYING_PREVIOUS_CYCLEWAY, isDisplayingPreviousCycleway)

    }

    private fun setAsResurvey(resurvey: Boolean) {
        isDisplayingPreviousCycleway = resurvey
        puzzleView.isEnabled = !resurvey
        if (resurvey) {
            setButtonsView(R.layout.quest_buttonpanel_yes_no)
            requireView().findViewById<View>(R.id.noButton).setOnClickListener {
                setAsResurvey(false)
            }
            requireView().findViewById<View>(R.id.yesButton).setOnClickListener {
                onClickOk()
            }
        } else {
            removeButtonsView()
        }
    }

    @AnyThread
    override fun onMapOrientation(rotation: Float, tilt: Float) {
        streetSideRotater?.onMapOrientation(rotation, tilt)
    }

    override fun onClickOk() {
        val leftSide = leftSide
        val rightSide = rightSide

        // a cycleway that goes into opposite direction of a oneway street needs special tagging
        var leftSideDir = 0
        var rightSideDir = 0
        var isOnewayNotForCyclists = false
        if (isOneway && leftSide != null && rightSide != null) {
            // if the road is oneway=-1, a cycleway that goes opposite to it would be cycleway:oneway=yes
            val reverseDir = if (isReversedOneway) 1 else -1

            if (isReverseSideRight) {
                if (rightSide.isSingleTrackOrLane()) {
                    rightSideDir = reverseDir
                }
            } else {
                if (leftSide.isSingleTrackOrLane()) {
                    leftSideDir = reverseDir
                }
            }

            isOnewayNotForCyclists = leftSide.isDualTrackOrLane() || rightSide.isDualTrackOrLane()
                    || (if(isReverseSideRight) rightSide else leftSide) !== Cycleway.NONE
        }

        applyAnswer(CyclewayAnswer(
            left = leftSide?.let { CyclewaySide(it, leftSideDir) },
            right = rightSide?.let { CyclewaySide(it, rightSideDir) },
            isOnewayNotForCyclists = isOnewayNotForCyclists
        ))
    }

    private fun Cycleway.isSingleTrackOrLane() =
        this === Cycleway.TRACK || this === Cycleway.EXCLUSIVE_LANE

    private fun Cycleway.isDualTrackOrLane() =
        this === Cycleway.DUAL_TRACK || this === Cycleway.DUAL_LANE

    override fun isFormComplete() = !isDisplayingPreviousCycleway && (
        if (isDefiningBothSides) leftSide != null && rightSide != null
        else                     leftSide != null || rightSide != null
    )

    override fun isRejectingClose() =
        !isDisplayingPreviousCycleway && (leftSide != null || rightSide != null)

    private fun showCyclewaySelectionDialog(isRight: Boolean) {
        val ctx = context ?: return
        val items = getCyclewayItems(isRight).map { it.asItem(isLeftHandTraffic) }
        ImageListPickerDialog(ctx, items, R.layout.labeled_icon_button_cell, 2) { selected ->
            val cycleway = selected.value!!
            val iconResId = cycleway.getIconResId(isLeftHandTraffic)
            val titleResId = resources.getString(cycleway.getTitleResId())

            if (isRight) {
                puzzleView.replaceRightSideImageResource(iconResId)
                puzzleView.setRightSideText(titleResId)
                rightSide = cycleway
            } else {
                puzzleView.replaceLeftSideImageResource(iconResId)
                puzzleView.setLeftSideText(titleResId)
                leftSide = cycleway
            }
            checkIsFormComplete()
        }.show()
    }

    private fun getCyclewayItems(isRight: Boolean): List<Cycleway> {
        val country = countryInfo.countryCode
        val values = DISPLAYED_CYCLEWAY_ITEMS.filter { it.isAvailableAsSelection(country) }.toMutableList()
        // different wording for a contraflow lane that is marked like a "shared" lane (just bicycle pictogram)
        if (isOneway && isReverseSideRight == isRight) {
            Collections.replaceAll(values, Cycleway.PICTOGRAMS, Cycleway.NONE_NO_ONEWAY)
        }
        return values
    }

    private fun showBothSides() {
        isDefiningBothSides = true
        puzzleView.showBothSides()
        checkIsFormComplete()
    }

    companion object {
        private const val CYCLEWAY_LEFT = "cycleway_left"
        private const val CYCLEWAY_RIGHT = "cycleway_right"
        private const val DEFINE_BOTH_SIDES = "define_both_sides"
        private const val IS_DISPLAYING_PREVIOUS_CYCLEWAY = "is_displaying_previous_cycleway"

    }
}
